tests/semantic_asserts.py

```diff
@@ -0,0 +1,299 @@
+            log the subroutine's result
+        * TealType.none or TealType.anytype:
+            log Itob(Int(1337)) as it is either impossible (TealType.none),
+            or unknown at compile time (TealType.any) how to convert to Bytes
+
+
+    Example 1: Using e2e_pyteal for a simple test of both an app and logic sig:
+        .. code-block:: python
+            from algosdk.testing.teal_blackbox import (
+                DryRunEncoder,
+                DryRunExecutor as Executor,
+            )
+
+            @Subroutine(TealType.uint64, input_types=[TealType.uint64])
+            def square(x):
+                return x ** Int(2)
+
+            # use this function to create pyteal app and signature expression approval functions:
+            approval_app = e2e_pyteal(square, Mode.Application)
+            approval_lsig = e2e_pyteal(square, Mode.Signature)
+
+            # compile the evaluated approvals to generate TEAL code
+            app_teal = compileTeal(approval_app(), Mode.Application, version=6)
+            lsig_teal = compileTeal(approval_lsig(), Mode.Signature, version=6)
+
+            # provide args for evaluation (will compute x^2)
+            x = 9
+            args = [x]
+
+            # evaluate the programs
+            algod = algod_with_assertion()
+            app_result = Executor.dryrun_app(algod, app_teal, args)
+            lsig_result = Executor.dryrun_logicsig(algod, lsig_teal, args)
+
+            # check to see that x^2 is at the top of the stack as expected
+            assert app_result.stack_top() == x ** 2, app_result.report(
+                args, "stack_top() failed for app"
+            )
+            assert lsig_result.stack_top() == x ** 2, lsig_result.report(
+                args, "stack_top() failed for lsig"
+            )
+
+            # check to see that itob of x^2 has been logged (only for the app case)
+            assert app_result.last_log() == DryRunEncoder.hex(x ** 2), app_result.report(
+                args, "last_log() failed for app"
+            )
+
+    Example 2: Using e2e_pyteal to make 400 assertions and generate a CSV report with 400 dryrun rows
+        .. code-block:: python
+            from itertools import product
+            import math
+            from pathlib import Path
+            import random
+
+            from algosdk.testing.teal_blackbox import (
+                DryRunExecutor as Executor,
+            )
+
+            @Subroutine(TealType.uint64, input_types=[TealType.uint64, TealType.uint64])
+            def euclid(x, y):
+                a = ScratchVar(TealType.uint64)
+                b = ScratchVar(TealType.uint64)
+                tmp = ScratchVar(TealType.uint64)
+                start = If(x < y, Seq(a.store(y), b.store(x)), Seq(a.store(x), b.store(y)))
+                cond = b.load() > Int(0)
+                step = Seq(
+                    tmp.store(b.load()), b.store(Mod(a.load(), b.load())), a.store(tmp.load())
+                )
```

Comment:
@ahangsu @jasonpaulos - is there a better way to write this for loop? I had to provide `Do()` with a dummy instruction `Pop(Int(1))` but I'd rather avoid this if possible.

Reply:
Probably an empty `Seq()`? Since `Do` wants an expression that has none type, then empty seq of instructions might be an option?

Reply:
👍

Reply:
It worked! Thank you @ahangsu